allenact/algorithms/onpolicy_sync/runner.py

```diff
@@ -1047,10 +1134,20 @@ def process_test_packages(
             log_writer.add_scalar(
                 f"{mode}-misc/num_tasks_evaled", num_tasks, training_steps
             )
+        callback_metric_means[f"{mode}-misc/num_tasks_evaled"] = num_tasks
 
         message.append(f"tasks {num_tasks} checkpoint {checkpoint_file_name[0]}")
         get_logger().info(" ".join(message))
 
+        for callback in self.callbacks:
```

Comment:
I guess it's also possible to make the callback start a thread and return immediately to allow logger to keep showing stats for train in "real time". Is that right?

Reply:
This would be possible, but one of the issues is that if wandb is initialized in one process, you cannot log from a thread, and hence why logging from a thread and then destroying it wouldn’t work as one would expect. If you try to log from all threads and processes, logging becomes prohibitively expensive.

See their notes on distributed training: https://docs.wandb.ai/guides/track/advanced/distributed-training

Reply:
I think I didn't express correctly what I meant. Currently, if the callback function takes a long time to process, packages in the runner queue for logging wait for a long time. When they're read, we quickly flush that queue, resulting in e.g. spikes of FPS (note that there's no timestamp in the logging packages sent from the trainers). I thought that passing the data for the callback to a thread (in the same process) could do the job, but if using an open wandb session from a thread in the same process is a no go, then there's no discussion. 👍